### DISCUSSION TITLE
How to enable over 20000 concurrent connection in windows (MQTT 3.1.1)

### DISCUSSION THREAD

**Discussion by zhitaohB** [rabbitmq/rabbitmq-server#4618] [Other] [answered]
I used  jmeter to test rabbitmq perforemance and found that 15000 concurrent connection works fine. But I get following error when testing 20000 concurrent connection

**_Caused by: io.netty.channel.AbstractChannel$AnnotatedSocketException: No buffer space available (maximum connections reached?): connect: /XXXX:20096_**
![下载 (1)](https://user-images.githubusercontent.com/102639279/164967222-cd4ddbcc-8a1d-4531-bb9c-64d6ae754654.png)

rabbitmq.config

```
mqtt.listeners.tcp.default = 1883
mqtt.listeners.tcp.1 = 20096
mqtt.proxy_protocol = false
mqtt.allow_anonymous = true
mqtt.vhost = /
mqtt.subscription_ttl = 1800000
mqtt.exchange = amq.topic
mqtt.prefetch = 10
mqtt.default_user = guest
mqtt.default_pass = guest
web_mqtt.tcp.port =  20072
loopback_users = none

mqtt.tcp_listen

_options.backlog = 128
mqtt.tcp_listen_options.nodelay = true
mqtt.tcp_listen_options.linger.on      = true
mqtt.tcp_listen_options.linger.timeout = 0
mqtt.tcp_listen_options.sndbuf  = 131072
mqtt.tcp_listen_options.recbuf  = 131072
```

![下载](https://user-images.githubusercontent.com/102639279/164967109-297b606c-524d-4746-8ab3-2cbc5cda6ee9.png)

I don't know what to do to enable over 20000 concurrent connections in windows (MQTT 3.1.1). 

Please give me some advice.

**Comment by michaelklishin** [comment, marked as answer]
See [Tuning for a Large Number of Connections](https://www.rabbitmq.com/networking.html#tuning-for-large-number-of-connections). Similar tuning may have to happen on the hosts that *open* those 20K connections.

While not directly related, RabbitMQ 3.8.2 is [out of support](https://www.rabbitmq.com/versions.html) and is 27 patch releases behind even in the 3.8 series alone.

**Reply by zhitaohB** [reply]
Thanks for your suggestions. I finally find my solution through this link https://confluence.atlassian.com/confkb/no-buffer-space-available-maximum-connections-reached-due-to-ephemeral-port-exhaustion-956142091.html

**Reply by michaelklishin** [reply]
Thank you for following up here. Client port exhaustion can be a problem, too, if you run all clients from a single machine.
